### PR TITLE
Return error when no error response

### PR DESCRIPTION
**Pull request by benmur**
This avoids silent failures on blob PUTs, as seen when proxying B2.

When the server closes the connection (timeouts on slow uploads, or keep-alive connections too old), causing an `IOException` (sorry, don't have the stack trace anymore) before the upload has finished, the error was not caught, making the client think that the upload went through.

502 seems to be an appropriate code for clients to go ahead and retry as expected - fixed the backup application I'm using.

### CODE REVIEW

**Comment by andrewgaul** [comment]
@benmur This seems reasonable.  Do you know a way to reproduce these symptoms?  I would like to see if there is a problem in the underlying jclouds library.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 1]
```diff
@@ -1722,6 +1722,7 @@ private static void handlePutBlob(HttpServletRequest request,
         } catch (HttpResponseException hre) {
             HttpResponse hr = hre.getResponse();
             if (hr == null) {
+                response.sendError(HttpServletResponse.SC_BAD_GATEWAY);
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Maybe we should log `hre` to discover the cause?  Also perhaps `INTERNAL_ERROR` is more appropriate until we understand the root cause?

**Comment by benmur** [comment]
The way I use to reproduce is upload about 1GB to B2 with restic (which embeds the minio-go s3 client) through S3Proxy. Every other time, the `IOException` occurs and is logged (directly by jclouds I think). I'll do an upload now to capture the jclouds stacktrace.

**Comment by benmur** [comment]
@andrewgaul branch updated with an error log.

My test triggered the issue early, Here is the logged hre:
```
[s3proxy] E 05-10 21:24:42.769 S3Proxy-36 o.gaul.s3proxy.S3ProxyHandler:1725 |::] Put blob failed without response
org.jclouds.http.HttpResponseException: Error writing request body to server connecting to POST https://pod-....backblaze.com/b2api/v1/b2_upload_file/... HTTP/1.1
        at org.jclouds.http.internal.BaseHttpCommandExecutorService.invoke(BaseHttpCommandExecutorService.java:122)
        at org.jclouds.rest.internal.InvokeHttpMethod.invoke(InvokeHttpMethod.java:90)
        at org.jclouds.rest.internal.InvokeHttpMethod.apply(InvokeHttpMethod.java:73)
        at org.jclouds.rest.internal.InvokeHttpMethod.apply(InvokeHttpMethod.java:44)
        at org.jclouds.reflect.FunctionalReflection$FunctionalInvocationHandler.handleInvocation(FunctionalReflection.java:117)
        at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:87)
        at com.sun.proxy.$Proxy58.uploadFile(Unknown Source)
        at org.jclouds.b2.blobstore.B2BlobStore.putBlob(B2BlobStore.java:269)
        at sun.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.google.inject.internal.DelegatingInvocationHandler.invoke(DelegatingInvocationHandler.java:37)
        at com.sun.proxy.$Proxy42.putBlob(Unknown Source)
        at org.gaul.s3proxy.S3ProxyHandler.handlePutBlob(S3ProxyHandler.java:1720)
        at org.gaul.s3proxy.S3ProxyHandler.doHandle(S3ProxyHandler.java:651)
        at org.gaul.s3proxy.S3ProxyHandlerJetty.handle(S3ProxyHandlerJetty.java:63)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.eclipse.jetty.server.Server.handle(Server.java:499)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:258)
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Error writing request body to server
        at sun.net.www.protocol.http.HttpURLConnection$StreamingOutputStream.checkError(HttpURLConnection.java:3518)
        at sun.net.www.protocol.http.HttpURLConnection$StreamingOutputStream.write(HttpURLConnection.java:3501)
        at com.google.common.io.CountingOutputStream.write(CountingOutputStream.java:53)
        at com.google.common.io.ByteStreams.copy(ByteStreams.java:179)
        at org.jclouds.http.internal.JavaUrlHttpCommandExecutorService.writePayloadToConnection(JavaUrlHttpCommandExecutorService.java:298)
        at org.jclouds.http.internal.JavaUrlHttpCommandExecutorService.convert(JavaUrlHttpCommandExecutorService.java:171)
        at org.jclouds.http.internal.JavaUrlHttpCommandExecutorService.convert(JavaUrlHttpCommandExecutorService.java:65)
        at org.jclouds.http.internal.BaseHttpCommandExecutorService.invoke(BaseHttpCommandExecutorService.java:99)
        ... 23 common frames omitted

```

Here are the jclouds-emitted log and stacktrace (slightly edited):
```
[s3proxy] E 05-10 21:24:42.766 S3Proxy-36 o.j.h.i.JavaUrlHttpCommandExecutorService:93 |::] error after writing 2157918/4295868 bytes to https://...
java.io.IOException: Error writing request body to server
        at sun.net.www.protocol.http.HttpURLConnection$StreamingOutputStream.checkError(HttpURLConnection.java:3518)
        at sun.net.www.protocol.http.HttpURLConnection$StreamingOutputStream.write(HttpURLConnection.java:3501)
        at com.google.common.io.CountingOutputStream.write(CountingOutputStream.java:53)
        at com.google.common.io.ByteStreams.copy(ByteStreams.java:179)
        at org.jclouds.http.internal.JavaUrlHttpCommandExecutorService.writePayloadToConnection(JavaUrlHttpCommandExecutorService.java:298)
        at org.jclouds.http.internal.JavaUrlHttpCommandExecutorService.convert(JavaUrlHttpCommandExecutorService.java:171)
        at org.jclouds.http.internal.JavaUrlHttpCommandExecutorService.convert(JavaUrlHttpCommandExecutorService.java:65)
        at org.jclouds.http.internal.BaseHttpCommandExecutorService.invoke(BaseHttpCommandExecutorService.java:99)
        at org.jclouds.rest.internal.InvokeHttpMethod.invoke(InvokeHttpMethod.java:90)
        at org.jclouds.rest.internal.InvokeHttpMethod.apply(InvokeHttpMethod.java:73)
        at org.jclouds.rest.internal.InvokeHttpMethod.apply(InvokeHttpMethod.java:44)
        at org.jclouds.reflect.FunctionalReflection$FunctionalInvocationHandler.handleInvocation(FunctionalReflection.java:117)
        at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:87)
        at com.sun.proxy.$Proxy58.uploadFile(Unknown Source)
        at org.jclouds.b2.blobstore.B2BlobStore.putBlob(B2BlobStore.java:269)
        at sun.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.google.inject.internal.DelegatingInvocationHandler.invoke(DelegatingInvocationHandler.java:37)
        at com.sun.proxy.$Proxy42.putBlob(Unknown Source)
        at org.gaul.s3proxy.S3ProxyHandler.handlePutBlob(S3ProxyHandler.java:1720)
        at org.gaul.s3proxy.S3ProxyHandler.doHandle(S3ProxyHandler.java:651)
        at org.gaul.s3proxy.S3ProxyHandlerJetty.handle(S3ProxyHandlerJetty.java:63)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.eclipse.jetty.server.Server.handle(Server.java:499)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:258)
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.lang.Thread.run(Thread.java:745)
[s3proxy] E 05-10 21:24:42.768 S3Proxy-36 o.j.h.h.BackoffLimitedRetryHandler:88 |::] Cannot retry after server error, command is not replayable: [method=org.jclouds.b2.features.ObjectApi.public abstract org.jclouds.b2.domain.UploadFileResponse org.jclouds.b2.features.ObjectApi.uploadFile(org.jclouds.b2.domain.UploadUrlResponse,java.lang.String,java.lang.String,java.util.Map,org.jclouds.io.Payload)[UploadUrlResponse{...}, restic/data/c8f748731cec021129e9ce07fcfc3e8167f4c70879e13927c86f2e37767a8e75, null, {}, [content=true, contentMetadata=[cacheControl=null, contentDisposition=null, contentEncoding=null, contentLanguage=null, contentLength=4295868, contentMD5=16db560d7566747a2acb469afb34894c, contentType=binary/octet-stream, expires=null], written=false, isSensitive=false]], request=POST https://... HTTP/1.1]

```

**Comment by andrewgaul** [comment]
@benmur Looking at this further, we should hoist the handling of `HttpResponseException` from `S3ProxyHandler.handlePutBlob` to `S3ProxyHandlerJetty.handle`.  This will require moving some of the digest logic from the former into the latter.  Could you look into this?  Thanks for working with me here!

**Comment by benmur** [comment]
@andrewgaul let me know if this is what you had in mind or if I missed something, then I'll cleanup the branch.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandlerJetty.java [hunk 2]
```diff
@@ -50,6 +54,13 @@
                 ignoreUnknownHeaders, corsAllowAll, servicePath);
     }
 
+    private void sendS3Exception(HttpServletRequest request,
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandlerJetty.java]: Add `static`.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandlerJetty.java [hunk 3]
```diff
@@ -69,10 +80,25 @@ public void handle(String target, Request baseRequest,
             baseRequest.setHandled(true);
             return;
         } catch (HttpResponseException hre) {
-            HttpResponse httpResponse = hre.getResponse();
-            response.sendError(httpResponse == null ?
-                    HttpServletResponse.SC_INTERNAL_SERVER_ERROR :
-                    httpResponse.getStatusCode());
+            HttpResponse hr = hre.getResponse();
+            if (hr == null) {
+                logger.error("Handler failed without response", hre);
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandlerJetty.java]: Remove log message.  Thanks to your earlier logging we know the server can just close the stream midway.

**Comment by benmur** [comment]
@andrewgaul thanks for the review, branch updated.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandlerJetty.java (hunk 2 above)
Comment (benmur) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandlerJetty.java]: Added `static`, although I have an afterthought: do we want `static` even though the method accesses `handler` which is an instance member?

Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandlerJetty.java]: Sorry I missed the access to `handler` -- agree that this looked better before my errant suggestion.

**Comment by benmur** [comment]
@andrewgaul updated, and I verified that with this fix the top-level error handler is called and returns a 500 error as expected.

**Comment by andrewgaul** [comment]
Thank you for your contribution @benmur!

**Pull request merged**